src/generator.rs

```diff
@@ -0,0 +1,140 @@
+use std::{
+    fmt,
+    path::{Path, PathBuf},
+};
+
+use eyre::Context;
+use serde::{Deserialize, Serialize};
+use tokio::fs;
+
+use crate::{
+    manifest::{Dependency, Manifest},
+    package::PackageStore,
+};
+
+/// The directory used for the generated code
+pub const BUILD_DIRECTORY: &str = "proto/build";
+
+/// The language used for code generation
+#[derive(
+    Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, clap::ValueEnum,
+)]
+#[serde(rename_all = "kebab-case")]
+pub enum Language {
+    Rust,
+}
+
+impl Language {
+    pub fn build_directory(&self) -> PathBuf {
+        Path::new(BUILD_DIRECTORY).join(self.to_string())
+    }
+}
+
+impl fmt::Display for Language {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", serde_typename::to_str(&self).unwrap_or("unknown"))
+    }
+}
+
+/// Backend used to generate code bindings
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Generator {
+    /// The tonic + prost stack
+    Tonic,
+}
+
+impl Generator {
+    pub const TONIC_INCLUDE_FILE: &str = "mod.rs";
+
+    /// Run the generator for a dependency and output files into `out`
+    pub async fn run(&self, dependency: &Dependency, out: &Path) -> eyre::Result<()> {
```

Comment:
curiosity - wouldn't you want to pass in all dependencies at once? if dependencies reference each other won't Tonic want access to all of those types? Otherwise I'd at the very least expect clashing definitions of Any etc.

Reply:
No! Buffrs sandboxes each dependency so if your deps look like:

```
you
  api-a@foo
    common@bar
  api-b@baz
    common@oof
```

You get:

```rust
mod api_a { mod common /* @bar for usage with api_a */ { .. } }
mod api_b { mod common /* @oof for usage with api_b */ { .. } }
```

Reply:
I can see why you've done this. I'm concerned that it leads to considerable ergonomic issues with regards to standard types - notably around conversions. For example, if APIs A and B both use some shared (and complex) type C, a consumer who reads a C from A and puts it into B will have to do conversions.
This seems to mitigate a large part of the advantage of being able to depend on shared protobuf libraries.

I'd wonder if there's some middle ground like Nix's 'follows' declarations? (you've used nix flakes a bunch, yeah?)

At the very least, it seems quite important to me that you don't use compile_well_known_types, so that at least Timestamp and Any don't have this issue.

Reply:
Upside is isolation, downside is that you can't use types of `api_a.common` with `api_b.common` w/o doing some conversion. I guess this a good tradeoff nevertheless though, but surprising coming from the status quo.

Reply:
> This seems to mitigate a large part of the advantage of being able to depend on shared protobuf libraries.

I agree, this may also change in the future. For now users are required to implement type conversions, as the actual dependency resolution would be a too complex problem to tackle as a side project

This is not saying we wont do it, but rather, we cant time wise do it in a sensible way

Reply:
I can disable the common type compilation 😊